unstructured/partition/text_type.py

```diff
@@ -57,6 +59,10 @@ def is_possible_narrative_text(
         If True, conducts checks that are specific to the chosen language. Turn on for more
         accurate partitioning and off for faster processing.
     """
+    if languages is None or languages == [""]:
+        languages = ["auto"]
+    if isinstance(languages, list) and "auto" in languages and text:
+        languages = detect_languages(text) or []
```

Comment:
@JIAQIA, this section needs to change. `detect_languages` should only be called when `detect_language_per_element=True` and with this change it is being called for almost every element.

`apply_lang_metadata` is the entry point all partitioners use for language detection and it determines if language detection is done at all. How we handle language detection admittedly needs to be refactored, but detecting the language of every element would slow down the processing of all documents, which we can't do.

If we just need to modify the condition for calling `exceeds_cap_ratio` to take into account languages that don't contain capitalization, we can modify this to just check the `language` parameter and you can set that when partitioning documents in other languages. Would work for you?

Reply:
@Coniferish Con The issue of detect_languages being called repeatedly is indeed a problem, and I’ve also noticed that it’s called again during apply_lang_metadata. However, as you mentioned, language detection is an area that needs deep refactoring. Given the current situation, undertaking such a deep refactor might be beyond what I can handle within a reasonable timeframe. Therefore, I chose the following solution, which is based on the following considerations:

1. If the user explicitly specifies the language as "eng", the repeated call to detect_language will be skipped. detect_language is only triggered when the user sets the language to auto or "" (an empty string). It’s worth noting that using "" to represent a specific meaning is hard to understand and process, and this should also be refactored.

2. To ensure that existing users are not affected, detect_language will not be called repeatedly if set language to "eng". If we want to enable this capability, I will specify language="auto" when using Unstructured. It’s important to note that if language="auto" is specified, then repeated calls to detect_language will indeed occur.

3. As I mentioned when I first proposed this solution, it will result in repeated calls to detect_languages. I believe the way to address this issue is to pass through detect_language_per_element during the partition process. Additionally, we need to consider the different behaviors when language="auto" and detect_language_per_element=False are configured, which should be illustrated using a Cartesian product to explain the program’s behavior under different combinations. If language detection has already been completed for an element in the first half of the partition process, this information should be stored in the element’s metadata. Furthermore, we need to optimize the behavior of apply_lang_metadata so that if the element’s metadata contains language information, the detection is skipped.

However, I am currently very limited on time, so I am unable to complete step 3. Step 3 would require extensive testing and coverage, and I believe that the extent of these changes would significantly increase the difficulty of merging.

Therefore, with my changes, the capabilities will only be activated when the user specifies that the document language is a non-English language like Chinese or "auto". If the user specifies "eng" or does not specify a language, my changes will not take effect.

Additionally, I hope the Unstructured team understands that metadata at the Element level is extremely important to us. If this metadata is incorrect, it is almost unusable in most scenarios. Therefore, we hope that this initial small step can be merged first, allowing non-English-speaking countries to use it before we discuss deep optimization.

To reiterate, this change ensures that when the user sets the language to "eng", the number of detect_language calls remains the same. But the number of detect_language calls will only change when the language belongs to the "languages without capitalization" list or **is set to "auto" (but sadlly, default value is auto)**, where there may be repeated calls (depending on the detect_language_per_element setting).

P.S. Regarding your suggestion to modify the conditions in the exceeds_cap_ratio function to account for languages that do not use capitalization, wouldn’t this also require calling detect_language? My opinion is that the root of this issue lies in the fact that the language parameter is not correctly passed during the partition process, and that the func call timing and logic of storing the results of detect_language are not ideal. So, based on my current understanding, modifying exceeds_cap_ratio might also introduce potential issues down the line.

Reply:
Hi @JIAQIA, I'm not proposing you to refactor the language behavior. That is something we would take care of ourselves, so please don't worry about that.

To your other points:
> 1. If the user explicitly specifies the language as "eng", the repeated call to detect_language will be skipped. detect_language is only triggered when the user sets the language to auto or "" (an empty string). It’s worth noting that using "" to represent a specific meaning is hard to understand and process, and this should also be refactored.

I agree that using an empty string is bad practice and something that needs to be changed. To clarify, though, it is being used to skip language detection by partitioners that depend on other partitioners (see [this comment](https://github.com/Unstructured-IO/unstructured/blob/1f8030dd0ee76389c619603197fc5f7f4e546d09/unstructured/partition/lang.py#L383))

> 2. To ensure that existing users are not affected, detect_language will not be called repeatedly if set language to "eng". If we want to enable this capability, I will specify language="auto" when using Unstructured. It’s important to note that if language="auto" is specified, then repeated calls to detect_language will indeed occur.

The default parameter for `languages` for all the document specific partitioners is `["auto"]`. Your changes would require users' behavior to change by having them passing `languages=["eng"]` if they do not want to detect the language per element.

When `language="auto"`, partitioning should not automatically detect language per element. That is the main point I am raising. There is a separate parameter, `detect_language_per_element` for controlling if language is detected per element and it should be the only parameter that has that responsibility. The `languages` parameter is provided for users to declare what language the document is in, mostly for the purposes of OCR and metadata, but it could also be used to detect the language before the element type is determined. I thought that the main issue you were trying to fix with this PR was the element type that was used for languages that do not use capitalization (reducing the number of `Title` and `UncategorizedText` elements that were detected and correctly categorizing them as `NarrativeText`). Please let me know if I'm mistaken about that.

I'm proposing that we keep your changes that make `languages` and `detect_language_per_element` accessible through the partition process. We could then use those parameters to more accurately determine if `exceeds_cap_ratio` needs to be called (which seemed to be the main change in determining the element type in this PR). Instead of relying on the default `languages=['auto']`, I'm saying that users partitioning non-English language documents could define the languages param (ex. `languages=['zho']`) and that could be used in `is_possible_narrative_text`. Or, users could leave the `languages` parameter alone and set `detect_language_per_element=True` and that could trigger `detect_languages` from within `is_possible_narrative_text`. 

> 3. As I mentioned when I first proposed this solution, it will result in repeated calls to detect_languages. I believe the way to address this issue is to pass through detect_language_per_element during the partition process. Additionally, we need to consider the different behaviors when language="auto" and detect_language_per_element=False are configured, which should be illustrated using a Cartesian product to explain the program’s behavior under different combinations. If language detection has already been completed for an element in the first half of the partition process, this information should be stored in the element’s metadata. Furthermore, we need to optimize the behavior of apply_lang_metadata so that if the element’s metadata contains language information, the detection is skipped.

Here is a table describing the current behavior:
| `detect_language_per_element` | `languages` | behavior |
|---|---|---|
| `True` | ["auto"] (default) |  Language is detected per element |
| `False` (default) | ["auto"] (default) |  Language is detected based on the text of the whole document |
| `True` | ["zho"] |  Language is detected per element (ignore `languages`) |
| `False` (default) | ["zho"] |  Language is not detected and "zho" is used |